technical-reports/format/composite-types.md

```diff
@@ -60,155 +75,205 @@ Other examples might be:
 
-- Groups are for arbitrarily grouping tokens for the purposes of naming and/or organization.
+- **[Groups](#groups)** are for arbitrarily grouping tokens for the purposes of naming and/or organization.
   - They impose no rules or restrictions on how many tokens or nested groups you put within them, what they are called, or what the types of the tokens within should be. As such, tools MUST NOT try to infer any special meaning or typing of tokens based on a group they happen to be in.
   - Different design systems are likely to group their tokens differently.
   - You can think of groups as containers that exist "outside" of design tokens.
-- Composite tokens are individual tokens whose value is made up of several sub-values following a predefined structure.
-  - The values of different composite tokens that share the same type are guaranteed to have the same internal structure (unlike multiple groups, where there is no guarantee whatsoever of how their contents are structured). Tools can therefore check their validity and potentially apply specialized processing or presentation to composite tokens.
-  - You can think of a composite type as something "inside" an individual token.
+- **Composite tokens** are individual design tokens whose values are made up of several sub-values.
+  - Since they are design tokens, they can be referenced by other design tokens. This is not true for groups.
+  - Their type must be one of the composite types defined in this specification. Therefore their names and types of their sub-values are pre-defined. Adding additional sub-values or setting values that don't have the correct type make the composite token invalid.
+  - Tools MAY provide specialised functionality for composite tokens. For example, a design tool may let the user pick from a list of all available shadow tokens when applying a drop shadow effect to an element.
+
+## Border
 
-## Type checking
+Represents a border style. The type property must be set to the string “border”. The value must be an object with the following properties:
 
-Just as with “normal” types for tokens, using a custom, composite type will allow tools to check that the values you use match the expected type. In our color pair example above, attempting to do the following would be invalid and tools should ignore the token and show an error to the user, since “Comic Sans MS” is not a valid color value.
+- `color`: The color of the border. The value of this property must be a valid [color value](#color) or a reference to a color token.
+- `width`: The width or thickness of the border. The value of this property must be a valid [dimension value](#dimension) or a reference to a dimension token.
+- `style`: The border's style, for example "solid" or "dashed". The value of this property must be a valid JSON string or a reference to a string token.
```

Comment:
Similar to font weight on the typography type, I wonder whether there should be a stand-alone `border-style` type that defines a set of known, permitted values (e.g. `"solid"`, `"dashed"`, `"dotted"`, etc.).

Otherwise any string would be valid and its unclear how tools should interpret the value

Reply:
I agree. Its value should be restricted to a set of the following strings: `solid`, `dashed`, `dotted`.

Reply:
Hmm... on second thoughts, a `borderStyle` (or perhaps `strokeStyle` is a better name?) type might need to be more complex. I've done a little bit of research and there's a few ways different platforms and tools do this:

* CSS has a finite number of ["line style" keywords](https://developer.mozilla.org/en-US/docs/Web/CSS/border-style#line-style) (`solid`, `dashed`, `dotted`, etc.), but doesn't offer finer control beyond that. 
* SVG has an attribute called [`stoke-dasharray`](https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/stroke-dasharray), which lets you define a patterns of dash and gap lengths. There's also a [`stroke-linecap`](https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/stroke-linecap) for round / butt / square caps.
* Figma has "Solid" / "Dash" / "Custom" styles. For dash and custom, you can set the dash & gap lengths in pretty much the same way as SVG. It also lets you choose from the same 3 linecap styles that SVG has, except they call it "Dash cap".
* Android seems to let you set [dash width and dash gap](https://developer.android.com/guide/topics/resources/drawable-resource#stroke-element) (so equivalent to Figma's "Dash" style), but doesn't support custom patterns like SVG & Figma do

This is of course not an exhaustive list (if anyone knows what iOS, Windows or other things support, then please share), but I think it's safe to say that a stroke style type is non-trivial.

Perhaps `strokeStyle` is itself a composite type? For example, we might define its values as objects with the following properties:

* `dashArray`: An array of `dimension` values, which could also be references to dimension tokens (as SVG does it)
* `lineCap`: One of the following strings `"round"`, `"butt"`, `"square"`

For example:

```json
{
  "line-dashed": {
    "type": "strokeStyle",
    "value": {
      "dashArray": ["3px", "{size-medium}"],
      "lineCap": "round"
    }
  },
  
  "size-medium": {
    "type": "dimension",
    "value": "5px"
  },
  
  "button-border": {
    "type": "border",
    "value": {
      "color": "#ff0000",
      "width": "{size-medium}",
      "style": "{line-dashed}"
    }
  }
}
```

But, similar to what I suggested for `fontWeight` in an earlier PR comment, we may _also_ want to have a set of keywords (similar to CSS) for common stroke styles. In that case, we ought to define some kind of mapping between the two. For example we might define `"dashed"` to be an alias for `{ "dashArray": ["3px", "3px"], "lineCap": "butt" }`. In that case, the following would also be valid:

```json
{
  "line-dashed": {
    "type": "strokeStyle",
    "value": "dashed"
  }
}  
```

One catch here is that the CSS line styles seem to scale relative to the border's width. So, the dashes and gaps on a `10px` border are wider than they are on a `1px` border. As far as I can tell, in SVG and Figma this is _not_ the case - the dash and gap values are essentially absolute lengths that remain the same regardless of how wide the border is. I threw together a little [CodePen to illustrate this](https://codepen.io/cirrus/pen/poWWBEO). At least in SVG, I couldn't figure out a way to make a dasharray scale relative to the stroke width. In other words SVG/Figma can do stuff CSS can't and CSS and do stuff that SVG/Figma can't! I have no idea what Android or iOS do in this respect (maybe @dbanksdesign knows?).

So maybe a mapping table isn't feasible?

-----

On a separate note, going down this route would also be setting an interesting precedent, since we'd have a token type that is both a composite and non-composite type in one! I can think of other cases where that might be useful. Do we need a name for that? "Hybrid type" perhaps? ;-)

Reply:
@c1rrus thanks for all the research you did on this. I think a `strokeStyle` composite type is needed, but should have the shorthand you're describing to align with CSS behaviors. 

I don't think we need a separate name for this type of token. It's still a composite token, it just has a shorthand for setting some of its sub values.

Reply:
+@kevinmpowell

Reply:
@kevinmpowell I've just pushed a commit that adds a new `stroke-style` composite type and updates the `border` type to use that.